drivers/battery/battc/battc.c

```diff
@@ -99,7 +101,14 @@ BatteryClassStatusNotify(PVOID ClassData)
     {
         case EVENT_BATTERY_TAG:
             ExReleaseFastMutex(&BattClass->Mutex);
-            DPRINT1("Waiting for battery is UNIMPLEMENTED!\n");
+            Status = BattClass->MiniportInfo.QueryTag(BattClass->MiniportInfo.Context,
+                                                      &Tag);
```

Comment:
What happens next to this (now assigned) Tag variable after the query?
Shouldn't you use the `BattWait->BatteryTag` member perhaps, instead?

Reply:
^ Also thought about that.

Reply:
**BATTERY_WAIT_STATUS** has nothing to do with what's the purpose of **EVENT_BATTERY_TAG** or with anything. `BattWait` is used as a hint for CMBATT that contains information under which circumstances should the battery status be gathered. Check MSDN and the rest of the battery implementation code for what it means.

The tag being queried here is to ensure, before we signal the event of which BATTC waits upon, that the composite battery REALLY has a tag assigned. And if not BATTC still has to wait until COMPBATT (or any miniport driver for that matter) gets a tag, and it can be queried, before BATTC completes the IRP and returns control to the kernel.

Reply:
Ok, so basically the code just wants to check "is there a tag assigned?" but doesn't care about its value (as long as it's not invalid) and will just signal the event.